docs/proposals/vision.md

```diff
@@ -0,0 +1,131 @@
+the stuff that the core Heapster/K8S team is responsible for and what is provided as a 
+plugin/addition and doesn’t come in the main release package. 
+
+We decided to focus only on:
+
+Kubernetes API Server node source
+Kubelet metrics source
+Influxdb, GCM, GKE (there is special endpoint for GKE that exposes all available metrics),
+ Hawkular sinks for Heapster
+Influxdb, GCL sinks for Eventer
+
+The rest of the sources/sinks will be available as plugins. The plugin will be used in 2 flavors:
+
+* Complied in - will require the user to rebuild the package and create his own image with 
+the desired set of plugins.  
+* Side-car - Heapster will talk to plugin’s HTTP server to get/pass metrics through a well 
+defined json interface. The plugin runs in a separate container.
+
+K8s team will explicitly say that it is NOT giving any warranty on the plugins. Plugins e2e 
+tests can be included in some CI suite but we will not block our development (too much) if 
+something breaks Kafka or Riemann. We will also not pay attention to whether a particular sink scales up.
+
+For now we will keep all of the currently available sinks compiled-in by default, to keep the 
+new Heapster more or less compatible with the old one, but eventually (if the number of sinks grows)
+ we will migrate some of them to plugins.
+
+## Custom Metrics Status
+
+Heapster is not a generic solution for gathering arbitrary number of arbitrary-formated custom 
+metrics. The support for custom metrics is focused on auto-scaling and critical functionality 
+monitoring (and potentially scheduling). And Heapster is oriented towards system metrics, not 
+application/business level metrics.
+
+Kubernetes users and application developers will be able to push any number of their custom 
+metrics through our pipeline to the storage but this should be considered as a bonus/best effort 
+functionality. Custom metrics will not influence our performance targets (no extra fine-tunning effort 
+to support >5 custom metrics per pod). There will be a flag in Kubelet that will limit the 
+number of custom metrics.
+
+## Performance Target
+
+Heapster product family (Core, Eventer and Oldtimer) should follow the same performance goals 
+as core Kubernetes. As Eventer is fairly simple and Oldtimer not yet fully defined this section
+will focus only on Core Heapster (for metrics).
+
+For 1.2 we should scale to 1000 nodes each running at least 30 pods (100 for 1.3) each reporting 
+20 metrics every 1 min (30 sec preferably). That brings us to the number of  600k metrics 
+per minute and 10k metrics per second.
+
+Stretch goal (for 1.2/1.3) is 60k metrics per second (possibly with not everything being written to Influxdb). 
+On smaller deployments, like 500 nodes with 15-30 pods each it should be easy to have 30 sec 
+metrics resolution or smaller. 
+
+Memory target - Fit into 2 gb with 1000 nodes x 30 pods and 6 gb with 1000 node x 100 pods (~60kb per pod). 
```

Comment:
Kindof. We don't want to use more than 60kb per pod (with 1 container) in total. In theory you can divide it by the number of samples you want to store and get some number (like 4kb for 15 samples for 15 minutes).

In our internal discussions we did it the other way round. We calculated the weight of all information and data structs needed to keep it and we got around 3.5 kb per pod sample.
While storing multiple samples of the same pod we can be smarter - we can store labels only once (if they didn't change) and reduce the memory needed. There also has to be some extra space reserved for parsing requests and building responses (and some/most formats are very inefficient - like storing all labels per metric/single value). So overall we are rather thinking about the whole system not about a single sample in isolation.